src/platforms/common/stm32/traceswoasync.c

```diff
@@ -106,35 +136,90 @@ void traceswo_setspeed(uint32_t baudrate)
 	usart_enable_rx_dma(SWO_UART);
 }
 
+#if defined(DMA_STREAM0)
 void SWO_DMA_ISR(void)
 {
-	if (DMA_ISR(SWO_DMA_BUS) & DMA_ISR_HTIF(SWO_DMA_CHAN)) {
-		DMA_IFCR(SWO_DMA_BUS) |= DMA_ISR_HTIF(SWO_DMA_CHAN);
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
 		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
 	}
-	if (DMA_ISR(SWO_DMA_BUS) & DMA_ISR_TCIF(SWO_DMA_CHAN)) {
-		DMA_IFCR(SWO_DMA_BUS) |= DMA_ISR_TCIF(SWO_DMA_CHAN);
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
 		memcpy(
 			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
 	}
 	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
 	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
 }
+#else
+void SWO_DMA_ISR(void)
+{
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
+		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
+	}
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
+		memcpy(
+			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
+	}
+	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
+	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
+}
+#endif
 
 void traceswo_init(uint32_t baudrate, uint32_t swo_chan_bitmask)
 {
+#ifdef TRACESWOASYNC_ALLOCS
+	if (trace_rx_buf == NULL) {
+		uint8_t *newbuf = memalign(TRACE_ENDPOINT_SIZE, NUM_TRACE_PACKETS * TRACE_ENDPOINT_SIZE);
+		if (newbuf == NULL) {
+			DEBUG_ERROR("%s: memalign failed, errno=%d\n", __func__, errno);
+			return;
+		}
+		trace_rx_buf = newbuf;
+	}
+#endif
+
 	if (!baudrate)
 		baudrate = SWO_DEFAULT_BAUD;
 
 	rcc_periph_clock_enable(SWO_UART_CLK);
 	rcc_periph_clock_enable(SWO_DMA_CLK);
 
+#if defined(STM32F1)
 	gpio_set_mode(SWO_UART_PORT, GPIO_MODE_INPUT, GPIO_CNF_INPUT_PULL_UPDOWN, SWO_UART_RX_PIN);
 	/* Pull SWO pin high to keep open SWO line ind uart idle state! */
 	gpio_set(SWO_UART_PORT, SWO_UART_RX_PIN);
+#else
+	gpio_mode_setup(SWO_UART_PORT, GPIO_MODE_AF, GPIO_PUPD_PULLUP, SWO_UART_RX_PIN);
+	gpio_set_output_options(SWO_UART_PORT, GPIO_OTYPE_OD, GPIO_OSPEED_100MHZ, SWO_UART_RX_PIN);
+	gpio_set_af(SWO_UART_PORT, GPIO_AF7, SWO_UART_RX_PIN);
+#endif
+
 	nvic_set_priority(SWO_DMA_IRQ, IRQ_PRI_SWO_DMA);
 	nvic_enable_irq(SWO_DMA_IRQ);
 	traceswo_setspeed(baudrate);
 	traceswo_setmask(swo_chan_bitmask);
 	decoding = (swo_chan_bitmask != 0);
 }
+
+void traceswo_deinit(void)
```

Comment:
What, if anything, currently calls this? If nothing, please include usage in this PR, such as a command in command.c that turns Trace/SWO capture off with this.

Reply:
Oops, nothing calls it. `command.c` should, but back then I ran out of determination to hook it into argv parser, which is already untrivial with all the parameters. The idea was to deinit on qRcmd upon stop or mode switch. I'll need some help on UI/naming. `monitor traceswo stop`? halt/deinit/reset/free?

Reply:
Given it's `mon traceswo` to enable it right now.. and that's not clear that it will actually enable decoding.. perhaps `enable` for the enable side, and `disable` for the deinit/stop? This would be consistent with other parts of BMD too then, including target power and such.

Reply:
Added extra command logic. I attempted to sidestep untangling the existing logic. Enabling in-probe ITM decode was optional and required the extra verbs (`mon traceswo 2250000 decode 2 3 30` or just `mon traceswo decode 2 3 30` for default baudrate). I think a real switch-case (or sequential ifs) parser is needed here for subcommands, especially for picking timer/uart modes.

Reply:
Agreed, which is something we have to tackle to get both back-ends compiling and working in the same build anyway - a long-term goal, but not something this PR needs to concern itself with.